src/pools/abstract-pool.ts

```diff
@@ -0,0 +1,172 @@
+import EventEmitter from 'events'
+import type { MessageValue } from '../utility-types'
+
+export type ErrorHandler<Worker> = (this: Worker, e: Error) => void
+export type OnlineHandler<Worker> = (this: Worker) => void
+export type ExitHandler<Worker> = (this: Worker, code: number) => void
+
+export interface PoolOptions<Worker> {
+  /**
+   * A function that will listen for error event on each worker.
+   */
+  errorHandler?: ErrorHandler<Worker>
+  /**
+   * A function that will listen for online event on each worker.
+   */
+  onlineHandler?: OnlineHandler<Worker>
+  /**
+   * A function that will listen for exit event on each worker.
+   */
+  exitHandler?: ExitHandler<Worker>
+  /**
+   * This is just to avoid not useful warnings message, is used to set `maxListeners` on event emitters (workers are event emitters).
+   *
+   * @default 1000
+   */
+  maxTasks?: number
+}
+
+interface IWorker {
+  on(event: 'error', handler: ErrorHandler<this>): void
+  on(event: 'online', handler: OnlineHandler<this>): void
+  on(event: 'exit', handler: ExitHandler<this>): void
+}
+
+class PoolEmitter extends EventEmitter {}
+
+export abstract class AbstractPool<
+  Worker extends IWorker,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  Data = any,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  Response = any
+> {
+  public readonly workers: Worker[] = []
+  public nextWorker: number = 0
+
+  /**
+   * `workerId` as key and an integer value
+   */
+  public readonly tasks: Map<Worker, number> = new Map<Worker, number>()
+
+  public readonly emitter: PoolEmitter
+
+  protected id: number = 0
+
+  public constructor (
+    public readonly numWorkers: number,
+    public readonly filePath: string,
+    public readonly opts: PoolOptions<Worker> = { maxTasks: 1000 }
+  ) {
+    if (!this.isMain()) {
+      throw new Error('Cannot start a pool from a worker!')
+    }
+    // TODO christopher 2021-02-07: Improve this check e.g. with a pattern or blank check
+    if (!this.filePath) {
+      throw new Error('Please specify a file with a worker implementation')
+    }
+
+    this.setupHook()
+
+    for (let i = 1; i <= this.numWorkers; i++) {
+      this.internalNewWorker()
+    }
+
+    this.emitter = new PoolEmitter()
+  }
+
+  protected setupHook (): void {
+    // Can be overridden
```

Comment:
This can be overriden or has to be overidden?
I mean if override is mandatory this should be an abstract method right?

Reply:
Like the comment say: You **can**  
So the function will called but it's default impl is just empty  
If you need to do something (like `setupMaster`) you can hook it up